mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp

```diff
@@ -201,6 +205,37 @@ class CmpFOpConversion : public OpConversionPattern<arith::CmpFOp> {
   }
 };
 
+/// Check if the signedness of type \p ty matches the expected
+/// signedness, and issue a type with the correct signedness if
+/// necessary.
+Type adaptIntegralTypeSignedness(Type ty, bool needsUnsigned) {
+  if (isa<IntegerType>(ty)) {
+    // Turns signless integers into signed integers.
+    if (ty.isUnsignedInteger() != needsUnsigned) {
+      auto signedness = needsUnsigned
+                            ? IntegerType::SignednessSemantics::Unsigned
+                            : IntegerType::SignednessSemantics::Signed;
+      return IntegerType::get(ty.getContext(), ty.getIntOrFloatBitWidth(),
+                              signedness);
+    }
+  } else if (emitc::isSizeTType(ty)) {
+    if (emitc::isUnsignedSizeTType(ty) != needsUnsigned) {
+      if (needsUnsigned)
+        return emitc::SizeTType::get(ty.getContext());
+      return emitc::SignedSizeTType::get(ty.getContext());
+    }
+  }
+  return ty;
+}
+
+/// Insert a cast operation to type \p ty if \p val
+/// does not have this type.
+Value adaptValueType(Value val, ConversionPatternRewriter &rewriter, Type ty) {
```

Comment:
This would be the same as `rewriter.createOrFold<emitc::CastOp>(val.getLoc(), ty, val);` if we had `emitc::CastOp::fold` that folded cast where srcTy == destTy.
So I suggest to either add that folder and use `rewriter.createOrFold`, or rename this function to `castOrFold`/`castIfNeeded`.

Reply:
Done, we're using the folder now. Didn't know about op folding, so I learnt something new!